README.md
# jojo [![Build status](https://travis-ci.org/twolfson/jojo.png?branch=master)](https://travis-ci.org/twolfson/jojo)
the 10 second blog-engine for hackers (in javascript).

It was built for running as middleware inside an [express][] app as well as a standalone application.

[express]: http://expressjs.com/2x/

```js
// articles/the-wonderful-wizard-of-oz.md
{
  "title": "The Wonderful Wizard of Oz",
  "author": "Lyman Frank Baum",
  "date": "1900/05/17"
}

Dorothy lived in the midst of the great Kansas prairies, with Uncle Henry,
who was a farmer, and Aunt Em, who was the farmer's wife.

// app.js
var jojo = require('jojo'),
    app = jojo();
app.listen(5000);
// Server is listening at http://localhost:5000/
// "The Wonderful Wizard of Oz" is being served at `/1900-05-16-the-wonderful-wizard-of-oz`
```

## Getting started
### Middleware
`jojo` is available via npm: `npm install jojo`

`jojo` is invoked with a configuration and can return a middleware or [express][] app.

```js
var app = jojo(); // express app
var app = jojo({render: false}); // vanilla middleware
```

By default, `jojo` reads articles that match `articles/*.{txt,md}` (`txt` and `md` files in `articles` directory) but this can be overridden via an option.

If you want more information on pattern options, see [node-glob][] reference.

[node-glob]: https://github.com/isaacs/node-glob

```js
var app = jojo({articles: 'papers/*.md'}); // Read articles from `papers`
```

By default, all articles are interpretted via [marked][], a [GitHub flavored markdown][gfm] parser.

More configuration options can be found in the [documentation][].

[marked]: https://github.com/chjj/marked
[gfm]: http://github.github.com/github-flavored-markdown/
[documentation]: #documentation

### Standalone
To run `jojo` as a standalone server, install `jojo` globally: `npm install -g jojo`.

`jojo` looks for a `jojo.js`/`jojo.json` to read in settings. These settings are passed to both a `jojo` middleware and [express][] app.

`jojo` also accepts arguments for choosing a different `config` or `port`.

```bash
$ jojo --help
Usage: jojo [options]
Options:
  -c, --config [config]  Config file to use (jojo.js, jojo.json)
  -p, --port [port]      Port to run server on (5000)
```

> `port` can also be specified in `jojo.js`/`jojo.json`

## Documentation
`jojo` has a wide spectrum of options


### jojo
`jojo` has an optional `options` parameter which modifies its behavior.

```js
jojo();
jojo(options);
/**
 * @param {Object} [options] Optional options to pass in to jojo
 * @returns {Function|Object} Based on options, returns
 *   a req/res/next function or express 2.x app.
 */
```

### options.render
`options.render` is expected as a `Boolean` and `true` by default.

When true, an [express 2.x][express] app will be returned by `jojo`.

Otherwise, a `req`/`res`/`next` middleware will be returned.

### options.articles
`options.articles` is expected as a `String` and `articles` by default.

This is the directory to read articles from. It is processed relative to `process.cwd()`.

### jojo.articles
`jojo.articles` is a sorted array of the parsed articles.

Each article has

- Any meta-data listed in the head of the file
- `rawContent`, unprocessed body
- `content`, formatted body
- `rawDate`, unprocessed date (from meta-data)
- `date`, `Date` object from `rawDate`
- `moment`, `moment` object from `rawDate` (uses timezone from `rawDate`)
- `url`, URL path for article. This can be modified via [options.urlFormatter][urlFormatter].
- `rawSummary`, if `summary` is provided, this will be the unprocessed `summary`
- `summary`, if `summary` is provided, this will be the formatted `summary`

### req.articles, res.locals.articles
Same as `jojo.articles` but located on `req` and `res` for easy accessibility.

### req.article, res.locals.article
On article pages, this will be the corresponding `article`.

See [urlFormatter][] for more `url` information

[urlFormatter]: #jojourlformatter

### req.config, res.locals.config
Configuration passed in to `jojo`. If invoked from standalone, this will be the as the `config`.

### jojo.dataParser
`jojo.dataParser` is a `Function` to separate meta data from the body of an article.

By default, this is [json-content-demux][]. If you are overriding, return an object containing meta data on `json` and body on `content`.

[json-content-demux]: https://github.com/twolfson/json-content-demux

### jojo.formatter
`jojo.formatter` is a `Function` to process the body into HTML markup.

By default, [marked][] is used. If you are overriding, return a string of HTML.

### jojo.urlFormatter
`jojo.urlFormatter` is a `Function` that receives an article and returns a URL path.

By default, it returns `/YYYY-MM-DD-article-name`. If you are overriding, return a string.

### jojo.articleSort
`jojo.articleSort` is a `Function` that compares articles to determine their order.

By default, it returns articles by most recent. If you are overriding, you will receive two articles and need to return an integer. See [Array.prototype.sort][] documentation for more information.

[Array.prototype.sort]: :https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort

## Donating
Support this project and [others by twolfson][twolfson-projects] via [donations][twolfson-support-me].

<http://twolfson.com/support-me>

[twolfson-projects]: http://twolfson.com/projects
[twolfson-support-me]: http://twolfson.com/support-me

## Attribution
`jojo` was initially inspired by [toto][].

[toto]: https://github.com/cloudhead/toto

## License
Copyright (c) 2013 Todd Wolfson

Licensed under the MIT license.


CHANGELOG.md
# jojo changelog
3.3.0 - Upgraded to request@2.88.0 to fix GitHub vulnerability warning

3.2.0 - Upgraded to express@3.21.2 to fix GitHub vulnerability warning

3.1.1 - Upgraded to latest Marked and Moment to resolve security issues

3.1.0 - Upgraded to latest EJS to resolve security issues

3.0.2 - Replaced Gratipay with support me page

3.0.1 - Added missing upgrade for `moment`

3.0.0 - Updated article dates to be based on timezone provided by date

2.0.0 - Added proper URL slugifying via `slugg`

1.1.2 - Moved from `doubleshot` to `mocha` due to `doubleshot` staleness

1.1.1 - Added `foundry` for releases

1.1.0 - Moved to glob for less greedy pattern matching. By default, articles must end in `.md` or `.txt`

1.0.0 - Moved to new API to statically load assets and simplify code.

0.2.0 - Added test suite and Travis CI. Scrapped old folder structure.

Before 0.1.7 - See `git log`


test/jojo_test.js
// Load in dependencies
var spawn = require('child_process').spawn,
    request = require('request'),
    expect = require('chai').expect;

var testUtils = {
  spawn: function (cmd, args, options) {
    before(function spawnChildProcess (done) {
      // Start up a child server
      this.child = spawn(cmd, args, options);

      // After it loads, callback
      setTimeout(done, 1000);
    });
    after(function killChildProcess (done) {
      // Kill the child process
      var child = this.child;
      delete this.child;
      child.kill();

      // When it leaves, callback
      child.on('exit', function (code) {
        done();
      });
    });
  },
  request: function (url) {
    before(function requestFn (done) {
      var that = this;
      request(url, function handleResponse (err, res, body) {
        // Save the error, response, and body
        that.err = err;
        that.res = res;
        that.body = body;

        // Callback with the error
        done(err);
      });
    });
    after(function cleanup () {
      delete this.err;
      delete this.res;
      delete this.body;
    });
  }
};

describe('jojo', function () {
  describe('run via CLI', function () {
    testUtils.spawn('node', ['../../../bin/jojo'], {
      cwd: __dirname + '/test_files/cli'
    });

    testUtils.request('http://localhost:11550/1900-05-17-the-wonderful-wizard-of-oz');
    it('serves articles', function () {
      expect(this.body).to.contain('<h1>The Wonderful Wizard of Oz</h1>');
    });
  });

  describe('integrated into a server', function () {
    testUtils.spawn('node', ['app.js'], {
      cwd: __dirname + '/test_files/integrated'
    });

    describe('with response to an index page', function () {
      testUtils.request('http://localhost:11550/');
      it('serves an index page', function () {
        expect(this.body).to.contain('<section id="articles">');
        expect(this.body).to.contain('the-wonderful-wizard-of-oz');
      });
    });

    describe('with respect to articles', function () {
      testUtils.request('http://localhost:11550/1900-05-17-the-wonderful-wizard-of-oz');
      it('serves articles', function () {
        expect(this.body).to.contain('<h1>The Wonderful Wizard of Oz</h1>');
      });
    });

    describe('with respect to a URL unfriendly article', function () {
      testUtils.request('http://localhost:11550/1900-05-17-plus-colon-hello');
      it('slugifies article titles', function () {
        expect(this.res.statusCode).to.equal(200);
      });
    });

    describe('with response to RSS', function () {
      testUtils.request('http://localhost:11550/index.xml');
      it('serves an RSS feed', function () {
        expect(this.body).to.contain('<id>/1900-05-17-the-wonderful-wizard-of-oz</id>');
      });
    });
  });
});

describe('jojo running with no articles', function () {
  testUtils.spawn('node', ['app.js'], {
    cwd: __dirname + '/test_files/empty_app'
  });

  testUtils.request('http://localhost:11550/');
  it('serves content', function () {
    expect(this.body).to.contain('<section id="articles">');
  });
});


src-test/jojo.test.js
{
  'jojo': {
    'can run standalone': '',
    'can integrate with a pre-existing server': '',
    'renders articles with text in the JSON': '',
    'renders article with text out of the JSON': '',
    'should generate proper xml': '',
      'even when there are no articles': ''
  }
}

test/jojo_test.content.js
// Load in dependencies
var spawn = require('child_process').spawn,
    request = require('request'),
    expect = require('chai').expect;

// Export doubleshot methods
module.exports = {
  // Assertions against jojo

  'serves an index page': [
    function requestHomepage () {
      this.url = 'http://localhost:11550/';
    }, 'makeRequest',
    function assertHomepage () {
      expect(this.body).to.contain('<section id="articles">');
      expect(this.body).to.contain('the-wonderful-wizard-of-oz');
    }
  ],
  'serves articles': [
    function requestArticle () {
      this.url = 'http://localhost:11550/1900-05-17-the-wonderful-wizard-of-oz';
    }, 'makeRequest',
    function assertArticle () {
      expect(this.body).to.contain('<h1>The Wonderful Wizard of Oz</h1>');
    }
  ],
  'serves an RSS feed': [
    function requestRss () {
      this.url = 'http://localhost:11550/index.xml';
    }, 'makeRequest',
    function assertRss () {
      expect(this.body).to.contain('<id>/1900-05-17-the-wonderful-wizard-of-oz</id>');
    }
  ],
  'serves content': [
    function requestEmptyHomepage () {
      this.url = 'http://localhost:11550/';
    }, 'makeRequest',
    function assertEmptyHomepage () {
      expect(this.body).to.contain('<section id="articles">');
    }
  ]
};


src/jojo.js
// Load in dependencies
var fs = require('fs'),
    url = require('url'),
    connect = require('connect'),
    glob = require('glob'),
    express = require('express'),
    _ = require('underscore'),
    jsonContentDemux = require('json-content-demux'),
    marked = require('marked'),
    moment = require('moment'),
    slugg = require('slugg');

// Class to handle storing articles
function ArticleCollection(options) {
  // Create placeholder for articles
  this._articles = [];

  // Fallback and save options for later
  this.options = options || {};
}
ArticleCollection.parseArticle = function (article, options) {
  // Fallback options
  options = options || {};

  // Demux content
  var demuxer = options.dataParser || jsonContentDemux,
      demuxedArticle = demuxer(article),
      data = demuxedArticle.json,
      markup = demuxedArticle.content;

  // Render the content
  var formatter = options.formatter || marked,
      content = formatter(markup);

  // Add information to data
  var retObj = data;
  retObj.rawDate = retObj.date;
  // https://github.com/moment/moment/blob/cd49a734350193a244c596a02167221cbca2c06f/moment.js#L1623-L1631
  // Parse moment with timezone attached to it
  retObj.moment = moment(retObj.date).parseZone();
  // DEV: Warning this date will be using server timezone
  retObj.date = retObj.moment.toDate();
  retObj.content = content;
  retObj.rawContent = markup;

  // If there is a summary, parse it via the formatter
  if (data.summary) {
    retObj.summary = formatter(data.summary);
  }

  // Add on defaults
  _.defaults(retObj, options.defaults || {});

  // Return the data
  return retObj;
};
ArticleCollection.prototype = {
  // Add a new article from plaintext
  addArticle: function (text) {
    // Parse and add the particle to our list
    var article = ArticleCollection.parseArticle(text, this.options);
    this._articles.push(article);
  },
  // Add a new article based on file path
  addArticleFromFile: function (filepath) {
    // Read and add the article
    var text = fs.readFileSync(filepath, 'utf8');
    this.addArticle(text, this.options);
  },
  // Return array of articles
  articles: function () {
    return this._articles;
  }
};

// Define our jojo middleware
function jojo(config) {
  // Fallback options
  config = config || {};

  // Create a new collection for our articles
  var collection = new ArticleCollection(config);

  // Locate and add articles from config
  var articlePattern = config.articles || process.cwd() + '/articles/*.{txt,md}',
      articleFiles = glob.sync(articlePattern);
  articleFiles.forEach(function addArticleFn (articlePath) {
    collection.addArticleFromFile(articlePath);
  });

  // Grab the articles
  var articles = collection.articles();

  // For each article, generate a URL
  var urlFormatter = config.urlFormatter || jojo.urlFormatter;
  articles.forEach(function generateUrl (article) {
    article.url = urlFormatter(article);
  });

  // Sort the articles
  var articleSort = config.articleSort || jojo.articleSort;
  articles.sort(articleSort);

  // Create a new server
  var app = connect();

  // Expose articles on the app
  app.articles = articles;

  // On all routes, expose articles and config
  app.use(function addArticles (req, res, next) {
    req.articles = articles;
    req.config = config;
    res.locals({
      articles: articles,
      config: config
    });
    next();
  });

  // Serve each article at its url
  articles.forEach(function serveArticles (article) {
    app.use(function serveArticleFn (req, res, next) {
      var parsedUrl = url.parse(req.url);
      if (parsedUrl.pathname === article.url) {
        req.article = article;
        res.locals({
          article: article
        });
      }
      next();
    });
  });

  // If we are supposed to render, then render
  var render = config.render === undefined ? true : config.render;
  if (render) {
    // Convert the app into an express app
    var jojoApp = app;
    app = express.createServer();
    app.use(jojoApp);

    // Render the homepage
    app.get('/', function renderIndex (req, res) {
      res.render('index');
    });

    // Render an XML webpage
    app.get('/index.xml', function renderRss (req, res) {
      res.render('xml');
    });

    // For each article, render it
    articles.forEach(function renderArticles (article) {
      app.get(article.url, function renderArticleFn (req, res) {
        res.render('article');
      });
    });
  }

  // Return the app
  return app;
}
// Default URL formatter, 1900-05-17-the-wonderful-wizard-of-oz
jojo.urlFormatter = function (article) {
  var dateStr = article.moment.format('YYYY-MM-DD'),
      url = dateStr + '-' + slugg(article.title);
  url = '/' + url.toLowerCase();
  return url;
};

// By default, sort by descending dates
jojo.articleSort = function (articleA, articleB) {
  return articleB.date - articleA.date;
};

// Expose ArticleCollection on jojo
jojo.ArticleCollection = ArticleCollection;

// Expose jojo
module.exports = jojo;
